### PR TITLE
Port to jbuilder take 2

### PR DESCRIPTION
**Pull request by rgrinberg**

### CODE REVIEW

**Comment by rgrinberg** [comment]
CI is passing now. @Drup have a look

**Comment by rgrinberg** [comment]
@Drup about to merge...

**Comment by Drup** [comment]
Did you tested the binary size differences in various scenarios ? Where were the results ?

File: re.opam [hunk 1]
```diff
@@ -1,34 +1,20 @@
...
 depends: [
-  "ocamlfind" {build}
-  "ocamlbuild" {build}
+  "jbuilder" {build}
   "base-bytes"
   "ounit" {test}
 ]
-
-# Modify in non-dev package.
-version: "dev"
```
Comment (Drup) [inline re.opam]: Please keep the dev version in the opam file.

Comment (rgrinberg) [inline re.opam]: Doesn't the topkg step substitute in the correct version for the pinned package anyway?

**Comment by rgrinberg** [comment]
> Did you tested the binary size differences in various scenarios ? Where were the results ?

I tested it as far as comparing the binary sizes of benchmarks/tests. The results were all completely identical. I can test it on external packages as well, but since jbuilder doesn't really differentiate between the two, there doesn't seem much use.

File: re.opam (hunk 1 above)
Comment (Drup) [inline re.opam]: I don't know how topkg works.

**Comment by Drup** [comment]
Good for me then. :)

**Comment by hcarty** [comment]
Keeping `dev` in the opam file makes `opam pin` do strange things with version constraints.  For example, `dev` ends up being less than `1.0.0` so it will violate version constraints for dependent packages.

**Comment by Drup** [comment]
This as come up multiple time. dev is above every other version in opam. If it's not the case, it's a bug and should be reported to @Altgr.

**Comment by avsm** [comment]
Is there anything blocking merging this?

**Comment by rgrinberg** [comment]
@avsm we need to do something else to maintain compat. I know how to do this, but we need drup on board.

**Comment by Drup** [comment]
Heh, if you keep compat, you have nothing to fear from me. :D

**Comment by rgrinberg** [comment]
OK but it requires reviving #148 and making some of the names ugly. We'd like to side-step the fact that jbuilder doesn't support cmi hiding by making use of module aliases. To preserve compat, we'll keep `re_{perl,pcre,emacs,str}` as their own libraries and make sure those keep working by aliasing them to `Re.{Perl,Pcre,Emacs,Str}`. The only downside, is that we'll have a few ugly type names leaking out. Note that all of the old type name will be aliases appropriately so nothing breaks.

@Drup you ok with that?

**Comment by Drup** [comment]
I forgot that ~~jbuilder~~ dune still doesn't support hiding cmis ... ugh, u_u'.

I don't really have the will to fight for this, since apparently people don't seem to care, so let's move on. 

Can you make a list of all the names that will be leaking, and see what we can do about them ? Also, rename Re0, for example to Core or something else slightly more evocative. :)
What do we lose if we don't use a wrapped library (except the fact that we would not be cool kids anymore ;) ?

**Comment by rgrinberg** [comment]
> I forgot that jbuilder dune still doesn't support hiding cmis ... ugh, u_u'.

Yes, though not for the lack of trying. It's just not as simple as not installing the cmi's. Dune is quite careful to treat external (findlib) and internal (defined with jbuild stanzas) libraries the same way. It's quite difficult to maintain this principle if you want to selectively hide cmi's. You basically need 2 build directories per 1 library. Quite yucky...

> Can you make a list of all the names that will be leaking, and see what we can do about them ? Also, rename Re0, for example to Core or something else slightly more evocative. :)

Sure. Note that this Core/Re0 will only be available to the user as `Re__{Core,Re0}`. We will re-export this module in the alias module itself.

It's a bit of a shame that dune will not let us have an `Re__Re` module. Because that would save us a bunch of renaming in all those `Re_{pcre,posix,..}` modules. Perhaps we can experiment with an alias like `module Re = Core`. I usually end up regretting tricks like this though. I'm particularly hesitant to do it for cosmetic reasons.

> What do we lose if we don't use a wrapped library (except the fact that we would not be cool kids anymore ;) ?

The wrapped library trick is equivalent to the cmi hiding trick to me. So the gain is having a mechanism to make modules private, even if it comes with an escape hatch. Another thing that is better with wrapping is that deprecate all those findlib sub libraries. I've found this to simplify things for new users. They expect a 1 to 1 correspondence between findlib and opam names (I know those expectations will be broken soon enough regardless, but it doesn't have to be by us).

**Comment by Drup** [comment]
> Sure. Note that this Core/Re0 will only be available to the user as Re__{Core,Re0}. We will re-export this module in the alias module itself.

Yes, I know, but still, I would prefer it to have a vaguely reasonable name.

> It's a bit of a shame that dune will not let us have an Re__Re module. Because that would save us a bunch of renaming in all those Re_{pcre,posix,..} modules. Perhaps we can experiment with an alias like module Re = Core. I usually end up regretting tricks like this though. I'm particularly hesitant to do it for cosmetic reasons.

I do not think that's necessary. also, if we have to choose between Re.Re or Re.Core, I'm going to go with the later. Afaics, the main "issue" we have is that Re.{Pcre,...} are going to refer to types of the form Re.Core.t in their API, is that right ? We might be able to do something about this using destructive substitution, but we lose module aliases ....

> The wrapped library trick is equivalent to the cmi hiding trick to me. [...] Another thing that is better with wrapping is that deprecate all those findlib sub libraries. [...]

Fair enough, let's do that.

**Pull request closed without merging**